code/modules/projectiles/guns/energy.dm

```diff
@@ -220,13 +223,64 @@
 			to_chat(user, SPAN_WARNING("You don't seem to know how to use [src]..."))
 			return FALSE
 
-/obj/item/weapon/gun/energy/taser/use_unique_action()
-	switch(precision)
-		if(TRUE)
-			precision = FALSE
-			to_chat(usr, SPAN_NOTICE("\The [src] is now set to Free mode."))
+/obj/item/weapon/gun/energy/taser/unique_action(mob/user)
+	change_mode(usr)
+
+/obj/item/weapon/gun/energy/taser/get_examine_text(mob/user)
+	. = ..()
+	switch(mode)
+		if(TASER_MODE_P)
+			. += SPAN_RED("It is set to precision mode, linked to the wanted database.")
+		if(TASER_MODE_F)
+			. += SPAN_GREEN("It is set to free mode, no longer linked to the wanted database.")
+
+/obj/item/weapon/gun/energy/taser/update_icon()
+	. = ..()
+	overlays += charge_icon + "_[mode]"
+
+/obj/item/weapon/gun/energy/taser/proc/change_mode(mob/user)
+	switch(mode)
+		if(TASER_MODE_P)
+			mode = TASER_MODE_F
+			to_chat(user, SPAN_NOTICE("\The [src] is now set to Free mode."))
 			ammo = GLOB.ammo_list[/datum/ammo/energy/taser]
-		if(FALSE)
-			precision = TRUE
-			to_chat(usr, SPAN_NOTICE("\The [src] is now set to Precision mode."))
+		if(TASER_MODE_F)
+			mode = TASER_MODE_P
+			to_chat(user, SPAN_NOTICE("\The [src] is now set to Precision mode."))
 			ammo = GLOB.ammo_list[/datum/ammo/energy/taser/precise]
+	var/datum/action/item_action/taser/change_mode/action = locate(/datum/action/item_action/taser/change_mode) in actions
```

Comment:
typecasting the var *might* make the locate type implicit? Would need to check this one